File: test/topology.cpp
#include <boost/ut.hpp>
#include <string_view>

#include "sys-sage.hpp"

using namespace boost::ut;
using namespace sys_sage;
using namespace std::string_view_literals;

static suite<"topology"> _ = []
{
    "Node"_test = []
    {
        Node node{42};
        expect(that % 42 == node.GetId());
        expect(that % ComponentType::Node == node.GetComponentType());
        expect(that % "Node"sv == node.GetComponentTypeStr());
        expect(that % nullptr == node.GetParent());

        Node root{0};
        node.SetParent(&root);
        expect(that % &root == node.GetParent());
    };

    "Topology"_test = []
    {
        Topology node;
        expect(that % 0 == node.GetId());
        expect(that % ComponentType::Topology == node.GetComponentType());
        expect(that % "Topology"sv == node.GetComponentTypeStr());
    };

    "Thread"_test = []
    {
        Node root{0};
        Thread node{&root, 42, "foo"};
        expect(that % &root == node.GetParent());
        expect(that % 42 == node.GetId());
        expect(that % "foo"sv == node.GetName());
        expect(that % ComponentType::Thread == node.GetComponentType());
        expect(that % "HW_Thread"sv == node.GetComponentTypeStr());
    };

    "Core"_test = []
    {
        Node root{0};
        Core node{&root, 42, "foo"};
        expect(that % &root == node.GetParent());
        expect(that % 42 == node.GetId());
        expect(that % "foo"sv == node.GetName());
        expect(that % ComponentType::Core == node.GetComponentType());
        expect(that % "Core"sv == node.GetComponentTypeStr());
    };

    "Cache"_test = []
    {
        Node root{0};
        Cache node{&root, 42, "3", 32, 2, 16};
        expect(that % &root == node.GetParent());
        expect(that % 42 == node.GetId());
        expect(that % "Cache"sv == node.GetName());
        expect(that % "3"sv == node.GetCacheName());
        expect(that % 3 == node.GetCacheLevel());
        expect(that % 32 == node.GetCacheSize());
        expect(that % 2 == node.GetCacheAssociativityWays());
        expect(that % 16 == node.GetCacheLineSize());
        expect(that % ComponentType::Cache == node.GetComponentType());
        expect(that % "Cache"sv == node.GetComponentTypeStr());

        node.SetCacheSize(16);
        expect(that % 16 == node.GetCacheSize());

        node.SetCacheLineSize(8);
        expect(that % 8 == node.GetCacheLineSize());
    };

    "Subdivision"_test = []
    {
        Node root{0};
        Subdivision node{&root, 42, "foo"};
        expect(that % &root == node.GetParent());
        expect(that % 42 == node.GetId());
        expect(that % "foo"sv == node.GetName());
        expect(that % ComponentType::Subdivision == node.GetComponentType());
        expect(that % "Subdivision"sv == node.GetComponentTypeStr());

        node.SetSubdivisionType(3);
        expect(that % 3 == node.GetSubdivisionType());
    };

    "Numa"_test = []
    {
        Node root{0};
        Numa node{&root, 42, 64};
        expect(that % &root == node.GetParent());
        expect(that % 42 == node.GetId());
        expect(that % "Numa"sv == node.GetName());
        expect(that % ComponentType::Numa == node.GetComponentType());
        expect(that % "NUMA"sv == node.GetComponentTypeStr());

        node.SetSubdivisionType(3);
        expect(that % 3 == node.GetSubdivisionType());
    };

    "Chip"_test = []
    {
        Node root{0};
        Chip node{&root, 42, "foo", 5};
        expect(that % &root == node.GetParent());
        expect(that % 42 == node.GetId());
        expect(that % "foo"sv == node.GetName());
        expect(that % ComponentType::Chip == node.GetComponentType());
        expect(that % "Chip"sv == node.GetComponentTypeStr());
        expect(that % 5 == node.GetChipType());

        node.SetModel("model");
        expect(that % "model"sv == node.GetModel());

        node.SetVendor("vendor");
        expect(that % "vendor"sv == node.GetVendor());

        node.SetChipType(6);
        expect(that % 6 == node.GetChipType());
    };

    "Memory"_test = []
    {
        Node root{0};
        Memory node{&root, 0, "foo", 32};
        expect(that % &root == node.GetParent());
        expect(that % 0 == node.GetId());
        expect(that % "foo"sv == node.GetName());
        expect(that % 32 == node.GetSize());
        expect(that % ComponentType::Memory == node.GetComponentType());
        expect(that % "Memory"sv == node.GetComponentTypeStr());

        node.SetSize(64);
        expect(that % 64 == node.GetSize());
    };

    "Storage"_test = []
    {
        Node root{0};
        Storage node{&root};
        expect(that % &root == node.GetParent());
        expect(that % ComponentType::Storage == node.GetComponentType());
        expect(that % "Storage"sv == node.GetComponentTypeStr());

        node.SetSize(64);
        expect(that % 64 == node.GetSize());
    };

    "Qubit"_test = []
    {
        Qubit q;
        q.SetProperties(1.0, 2.0, 3.0, 4.0, 5.0);
        expect(that % q.GetT1() == 1.0);
        expect(that % q.GetT2() == 2.0);
        expect(that % q.GetReadoutFidelity() == 3.0);
        expect(that % q.Get1QFidelity() == 4.0);
        expect(that % q.GetReadoutLength() == 5.0);
        expect(that % "Qubit"sv == q.GetComponentTypeStr());
    };

    "QuantumBackend"_test = []
    {
        QuantumBackend qb;
        size_t s = 3;
        QuantumGate gates[s];
        for (size_t i = 0; i < s; i++)
          qb.addGate(gates + i);

        expect(that % qb.GetNumberofGates() == s);

        std::vector<QuantumGate *> qg = qb.GetAllGateTypes();
        for (size_t i = 0; i < s; i++)
          expect(that % qg[i] == gates + i);

        expect(that % qb.GetGatesBySize(0) == qg);
        expect(that % "QuantumBackend"sv == qb.GetComponentTypeStr());
    };

    "Children insertion and removal"_test = []
    {
        Node a;
        Node b;
        Node c;
        Node d;

        a.InsertChild(&b);
        a.InsertChild(&c);
        a.InsertChild(&d);
        expect(that % 3_u == a.GetChildren().size());

        expect(that % 1 == a.RemoveChild(&b));
        expect(that % (2_u == a.GetChildren().size()) >> fatal);
        expect(that % (std::find(a.GetChildren().begin(), a.GetChildren().end(), &b) == a.GetChildren().end()));
        expect(that % (std::find(a.GetChildren().begin(), a.GetChildren().end(), &c) != a.GetChildren().end()));
        expect(that % (std::find(a.GetChildren().begin(), a.GetChildren().end(), &d) != a.GetChildren().end()));
    };

    "Get child"_test = []
    {
        Node a{1};
        Node b{2};
        Node c{3};
        Node d{4};

        a.InsertChild(&b);
        a.InsertChild(&c);
        b.InsertChild(&d);

        expect(that % a.GetChild(2) == &b);
        expect(that % a.GetChild(3) == &c);
        expect(that % a.GetChild(4) == nullptr);
    };

    "Get child by type"_test = []
    {
        Node a;
        Memory b;
        Memory c;
        Chip d;

        a.InsertChild(&b);
        a.InsertChild(&c);
        a.InsertChild(&d);

        expect(that % a.GetChildByType(ComponentType::Memory) == &b);
        expect(that % a.GetChildByType(ComponentType::Chip) == &d);
        expect(that % a.GetAllChildrenByType(ComponentType::Memory) == (std::vector<Component *>{&b, &c}));
    };

    "Get parent by type"_test = []
    {
        Cache a;
        Core b{&a};
        Thread c{&b};
        expect(that % &a == c.GetAncestorByType(ComponentType::Cache));
    };

    "Component tree consistency"_test = []
    {
        {
            Node a;
            Node b;
            Node c;
            a.InsertChild(&b);
            b.InsertChild(&c);
            expect(that % 0 == a.CheckComponentTreeConsistency());
        }
        {
            Node a;
            Node b;
            Node c;
            a.InsertChild(&b);
            b.InsertChild(&c);
            c.SetParent(&a);
            expect(that % 1 == a.CheckComponentTreeConsistency());
        }
    };

    "Get deeper components"_test = []
    {
        Node a;
        Node b;
        Node c;
        Node d;

        a.InsertChild(&b);
        a.InsertChild(&c);
        c.InsertChild(&d);

        std::vector<Component *> array;
        a.GetNthDescendents(&array, 1);
        expect(that % 2_u == array.size());
    };

    "Get subcomponents by type"_test = []
    {
        Node a;
        Chip b;
        Memory c;
        Chip d;

        a.InsertChild(&b);
        a.InsertChild(&c);
        c.InsertChild(&d);

        std::vector<Component *> array;
        a.GetSubcomponentsByType(&array, ComponentType::Chip);
        expect(that % 2_u == array.size());
    };

    "Get total number of threads"_test = []
    {
        Node a;
        Thread b;
        Thread c;
        Node d;
        Thread e;
        Node f;

        a.InsertChild(&b);
        a.InsertChild(&c);
        a.InsertChild(&d);
        d.InsertChild(&e);
        d.InsertChild(&f);

        expect(that % 3 == a.CountAllSubcomponentsByType(ComponentType::Thread));
    };

    "Linearize subtree"_test = []
    {
        Node a;
        Node b;
        Node c;
        Node d;

        a.InsertChild(&b);
        b.InsertChild(&d);
        a.InsertChild(&c);

        std::vector<Component *> array;
        a.GetComponentsInSubtree(&array);
        expect(that % array == (std::vector<Component *>{&a, &b, &d, &c}));
    };

    "Tree depth"_test = []
    {
        Node a;
        Node b;
        Node c;
        Node d;
        Node e;
        Node f;
        Node g;

        a.InsertChild(&b);
        b.InsertChild(&c);
        a.InsertChild(&d);
        d.InsertChild(&e);
        e.InsertChild(&f);
        a.InsertChild(&g);

        expect(that % 3 == a.GetSubtreeDepth());
    };
};

File: test/topology.py
import unittest
import py_sys_sage as pysage 

class TestComponents(unittest.TestCase):
    def test_node(self):
        n = pysage.Node(42)
        self.assertEqual(n.id, 42)
        self.assertEqual(n.type, pysage.COMPONENT_NODE)
        self.assertEqual(n.GetComponentTypeStr(), "Node")
        self.assertEqual(n.parent, None)
        
        root = pysage.Node(1)
        n.parent = root
        self.assertEqual(n.parent, root)
        
    def test_topologogy(self):
        t = pysage.Topology()
        self.assertEqual(t.id, 0)
        self.assertEqual(t.type, pysage.COMPONENT_TOPOLOGY)
        self.assertEqual(t.GetComponentTypeStr(), "Topology")
        
    def test_thread(self):
        root = pysage.Node(0)
        t = pysage.Thread(root, 42, "foo")
        self.assertEqual(root, t.parent)
        self.assertEqual(t.id, 42)
        self.assertEqual(t.type, pysage.COMPONENT_THREAD)
        self.assertEqual(t.GetComponentTypeStr(), "HW_Thread")
        self.assertEqual(t.name, "foo")
    
    def test_core(self):
        root = pysage.Node(0)
        c = pysage.Core(root, 42, "foo")
        self.assertEqual(root, c.parent)
        self.assertEqual(c.id, 42)
        self.assertEqual(c.type, pysage.COMPONENT_CORE)
        self.assertEqual(c.GetComponentTypeStr(), "Core")
        self.assertEqual(c.name, "foo")
        
    def test_cache(self):
        root = pysage.Node(0)
        c = pysage.Cache(root, 42, "3", 32, 2, 16)
        self.assertEqual(root, c.parent)
        self.assertEqual(c.id, 42)
        self.assertEqual(c.type, pysage.COMPONENT_CACHE)
        self.assertEqual(c.GetComponentTypeStr(), "Cache")
        self.assertEqual(c.name, "Cache")
        self.assertEqual(c.cacheName, "3")
        self.assertEqual(c.cacheLevel, 3)
        self.assertEqual(c.cacheSize, 32)
        self.assertEqual(c.cacheAssociativity, 2)
        self.assertEqual(c.cacheLineSize, 16)
        
    def test_subdivision(self):
        root = pysage.Node(0)
        s = pysage.Subdivision(root, 42, "foo")
        self.assertEqual(root, s.parent)
        self.assertEqual(s.id, 42)
        self.assertEqual(s.type, pysage.COMPONENT_SUBDIVISION)
        self.assertEqual(s.GetComponentTypeStr(), "Subdivision")
        self.assertEqual(s.name, "foo")
        
        s.subdivisionType = 3   
        self.assertEqual(s.subdivisionType, 3)
        
    def test_numa(self):
        root = pysage.Node(0)
        n = pysage.Numa(root, 42, 64)
        self.assertEqual(root, n.parent)
        self.assertEqual(n.id, 42)
        self.assertEqual(n.type, pysage.COMPONENT_NUMA)
        self.assertEqual(n.GetComponentTypeStr(), "NUMA")
        self.assertEqual(n.name, "Numa")
        
        n.subdivisionType = 3   
        self.assertEqual(n.subdivisionType, 3)
    
    def test_chip(self):
        root = pysage.Node(0)
        c = pysage.Chip(root, 42, "foo", 5)
        self.assertEqual(root, c.parent)
        self.assertEqual(c.id, 42)
        self.assertEqual(c.type, pysage.COMPONENT_CHIP)
        self.assertEqual(c.GetComponentTypeStr(), "Chip")
        self.assertEqual(c.name, "foo")
        self.assertEqual(c.chipType, 5)
        
        c.model = "model"
        self.assertEqual(c.model, "model")
        
        c.vendor = "vendor"
        self.assertEqual(c.vendor, "vendor")
        
        c.chipType = 6
        self.assertEqual(c.chipType, 6)
        
    def test_memory(self):
        root = pysage.Node(0)
        m = pysage.Memory(root, 0, "foo", 32)
        self.assertEqual(root, m.parent)
        self.assertEqual(m.id, 0)
        self.assertEqual(m.type, pysage.COMPONENT_MEMORY)
        self.assertEqual(m.GetComponentTypeStr(), "Memory")
        self.assertEqual(m.name, "foo")
        self.assertEqual(m.size, 32)
        
        m.size = 64
        self.assertEqual(m.size, 64)
    
    def test_storage(self):
        root = pysage.Node(0)
        s = pysage.Storage(root)
        self.assertEqual(root, s.parent)
        self.assertEqual(s.type, pysage.COMPONENT_STORAGE)
        self.assertEqual(s.GetComponentTypeStr(), "Storage")
        
        s.size = 64
        self.assertEqual(s.size, 64)    

    def test_qubit(self):
        q = pysage.Qubit()
        q.SetProperties(1.0, 2.0, 3.0, 4.0, 5.0);
        self.assertEqual(q.t1, 1.0)
        self.assertEqual(q.t2, 2.0)
        self.assertEqual(q.readout_fidelity, 3.0);
        self.assertEqual(q.q1_fidelity, 4.0);
        self.assertEqual(q.readout_length, 5.0);
        self.assertEqual(q.GetComponentTypeStr(), "Qubit")

    def test_quantum_backend(self):
        qb = pysage.QuantumBackend();
        s = 3
        gates = [pysage.QuantumGate()] * s
        for gate in gates:
            qb.addGate(gate)

        self.assertEqual(qb.GetNumberofGates(), s)
        self.assertEqual(qb.GetAllGateTypes(), gates)
        self.assertEqual(qb.FindGatesBySize(0), gates)
        self.assertEqual(qb.GetComponentTypeStr(), "QuantumBackend")

    def test_atom_site(self):
        a = pysage.AtomSite()
        # test if we can access the struct
        a.properties.nRows = 1
        a.properties.nColumns = 2
        a.properties.nAods = 3
        a.properties.nAodIntermediateLevels = 4
        a.properties.nAodCoordinates = 5
        a.properties.interQubitDistance = 6.0
        a.properties.interactionRadius = 7.0
        a.properties.blockingFactor = 8.0
        
    def test_children_insertion_and_removal(self):
        a = pysage.Node()
        b = pysage.Node()
        c = pysage.Node()
        d = pysage.Node()
        a.InsertChild(b)
        a.InsertChild(c)
        a.InsertChild(d)
        self.assertEqual(len(a.GetChildren()), 3)
        self.assertEqual(a.RemoveChild(b), 1)
        self.assertEqual(len(a.GetChildren()), 2)
        self.assertNotIn(b, a.GetChildren())
        self.assertIn(c, a.GetChildren())
        self.assertIn(d, a.GetChildren())
    
    def test_get_child(self):
        a = pysage.Node()
        b = pysage.Node(0)
        c = pysage.Node(1)
        d = pysage.Node(2)
        a.InsertChild(b)
        a.InsertChild(c)
        a.InsertChild(d)
        self.assertEqual(a.GetChild(0), b)
        self.assertEqual(a.GetChild(1), c)
        self.assertEqual(a.GetChild(4), None)
        
    def test_get_parent_by_type(self):
        a = pysage.Cache()
        b = pysage.Core(a)
        c = pysage.Thread(b)
    def test_component_tree_consistency(self):
        a = pysage.Node()
        b = pysage.Node()
        c = pysage.Node()
        a.InsertChild(b)
        b.InsertChild(c)
        
        self.assertEqual(0, a.CheckSubtreeConsistency())
        
        c.SetParent(a)
        self.assertEqual(1, a.CheckSubtreeConsistency())
        
    def test_get_deeper_components(self):
        a = pysage.Node()
        b = pysage.Node()
        c = pysage.Node()
        d = pysage.Node()
        a.InsertChild(b)
        a.InsertChild(c)
        c.InsertChild(d)
        array = []
        array += a.FindNthDescendants(1)
        self.assertEqual(2, len(array))
        
    def test_get_subcomponents_by_type(self):
        a = pysage.Node()
        b = pysage.Chip()
        c = pysage.Memory()
        d = pysage.Chip()
        
        a.InsertChild(b)
        a.InsertChild(c)
        c.InsertChild(d)
        
        array = []
        array += a.FindDescendantsByType(pysage.COMPONENT_CHIP)
        
        self.assertEqual(2, len(array))
    
    def test_get_total_number_of_threads(self):
        a = pysage.Node()
        b = pysage.Thread()
        c = pysage.Thread()
        d = pysage.Node()
        e = pysage.Thread()
        f = pysage.Node()
        
        a.InsertChild(b)
        a.InsertChild(c)
        a.InsertChild(d)
        d.InsertChild(e)
        d.InsertChild(f)
        
        self.assertEqual(3, a.CountDescendantsByType(pysage.COMPONENT_THREAD))
    
    def test_linearize_subtree(self):
        a = pysage.Node()
        b = pysage.Node()
        c = pysage.Node()
        d = pysage.Node()
        
        a.InsertChild(b)
        b.InsertChild(d)
        a.InsertChild(c)
        
        array = []
        array += a.FindDescendantsByType(pysage.COMPONENT_ANY)
        
        self.assertEqual(array, [a, b, d, c])
        
    def test_tree_depth(self):
        a = pysage.Node()
        b = pysage.Node()
        c = pysage.Node()
        d = pysage.Node()
        e = pysage.Node()
        f = pysage.Node()
        g = pysage.Node()
        
        a.InsertChild(b)
        b.InsertChild(c)
        a.InsertChild(d)
        d.InsertChild(e)
        e.InsertChild(f)
        a.InsertChild(g)
        
        self.assertEqual(3, a.CalcSubtreeDepth())

    def test_attrib(self):
        c = pysage.Component()

        c.SetAttribute("foo", 1)
        c.SetAttribute("bar", 2.0)
        c.SetAttribute("foobar", "test")

        self.assertEqual(c.GetAttribute("foo"), 1)
        self.assertEqual(c.GetAttribute("bar"), 2.0)
        self.assertEqual(c.GetAttribute("foobar"), "test")

        c.EraseAttribute("foo")
        self.assertEqual(c.GetAttribute("foo"), None)

        c.ClearAttributes()
        self.assertEqual(c.GetAttributesSize(), 0)

if __name__ == "__main__":
    unittest.main()